back/flag-service/src/flag/user/spec/UserController.spec.ts

```diff
@@ -0,0 +1,340 @@
+import { BadRequestException } from "@nestjs/common";
+import { UserController } from "../UserController";
+import { UserService } from "../UserService";
+import { EmailAlreadyTakenError } from "../errors/EmailAlreadyTakenError";
+import { NicknameAlreadyTakenError } from "../errors/NicknameAlreadyTakenError";
+import { EmailNotFoundError } from "../errors/EmailNotFoundError";
+import { IncorrectPasswordError } from "../errors/IncorrectPasswordError";
+
+
+describe('UserController', () => {
+  let userController: UserController;
+  let userService: UserService;
+
+  beforeAll(async () => {
+    userService = new UserService(null);
+    userController = new UserController(userService);
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+
+  describe('register', () => {
+    describe('success', () => {
+      let registerSpy;
+      let res;
+      beforeAll(async () => {
+        registerSpy = jest
+          .spyOn(userService, 'register')
+          .mockReturnValue({ fake: true } as any);
+        res = await userController.register('user@example.com', 'secret', 'secret', 'jane');
+      });
+      it('calls register from service', () => {
+        expect(registerSpy).toBeCalledTimes(1);
+      });
+      it('returns register return value', () => {
+        expect(res).toStrictEqual({ success: true });
+      });
+    });
+    describe('failure', () => {
+      for (const error of [
+        new EmailAlreadyTakenError(),
+        new NicknameAlreadyTakenError(),
+      ]) {
+        describe(error.constructor.name, () => {
+          let registerSpy;
+          let res;
+          beforeAll(async () => {
+            registerSpy = jest
+              .spyOn(userService, 'register')
+              .mockImplementation(() => {
+                throw error;
+              });
+            res = userController.register('user@example.com', 'thisisasecret123', 'thisisasecret123', 'jane');
+          });
+          it('calls register from service', () => {
+            expect(registerSpy).toBeCalledTimes(1);
+          });
+          it('throws a BadRequestException when service throws a ' + error.constructor.name, async () => {
+            await expect(res).rejects.toThrow(BadRequestException);
+          });
+        });
+      }
+
+      describe('bad field values', () => {
+        function testValues(name: string, email: string, password: string, passwordConfirmation: string, nickname: string) {
```

Comment:
maybe rename the function to `testBadValues` or something else that convey the idea we are testing faulty values